FastAPI: Separate model server and data plane

sukumargaonkar opened the pull request:
**What this PR does / why we need it**:
The main objective is to migrate the kserve sdk from using tornado to fastapi
As first step this PR encapsulates the kserve sdk in a dataplane api which would later be used as building blocks to implement FastApi endpoints

fixes #2479 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

yuzisun committed:
Update kserve  python dependency

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun committed:
Convert to FastAPI

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun committed:
Convert to FastAPI

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun committed:
Add workers

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

sukumargaonkar committed:
Add Dataplane api and model repo extension api

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun commented on docs/samples/kafka/setup.py:
```diff
@@ -30,7 +30,7 @@
     packages=find_packages("transformer"),
     install_requires=[
-        "kserve",
+        "kserve>=0.9.0",
```
wondering if there are better way to set the version, otherwise we need to update many places every time for the release.

yuzisun commented on python/kserve/kserve/handlers/dataplane.py:
```diff
@@ -0,0 +1,246 @@
+    HTTP Servers should return HTTP_404 (Not Found).
+    """
+
+    def __init__(self, model_name=None):
```
any reason to default to None here?

yuzisun commented on python/kserve/kserve/model.py:
```diff
@@ -92,7 +103,7 @@ def __init__(self, name: str):
 
     async def __call__(self, body, model_type: ModelType = ModelType.PREDICTOR, headers: Dict[str, str] = None):
-        request_id = headers.get("X-Request-Id")
+        request_id = headers.get("X-Request-Id", "N.A.") if headers else "N.A."
```
should we generate the Id if it is not there?

yuzisun commented on python/kserve/kserve/handlers/__init__.py:
```diff
@@ -17,3 +17,4 @@
 from .model_management import LoadHandler, UnloadHandler, ListHandler  # noqa # pylint: disable=unused-import
 from .explain import ExplainHandler  # noqa # pylint: disable=unused-import
 from .predict import PredictHandler  # noqa # pylint: disable=unused-import
+from .dataplane import DataPlane
```
nit: newline

yuzisun commented on python/kserve/kserve/handlers/dataplane.py:
```diff
@@ -0,0 +1,246 @@
+            body = self.get_binary_cloudevent(body, headers)
+        else:
+            try:
+                body = json.loads(body)
```
we might want to switch to use `rapidjson` for better serialization/deserialization performance especially for large inputs
https://pythonspeed.com/articles/faster-json-library/

yuzisun commented on python/kserve/kserve/handlers/dataplane.py:
```diff
@@ -0,0 +1,246 @@
+                raise InvalidInput(f"Unrecognized request format: {e}")
+
+            if is_structured_cloudevent(body):
+                is_cloudevent = True
```
Can these logic be abstracted into decode a func?

yuzisun commented on python/kserve/kserve/model_server.py:
```diff
@@ -79,39 +74,24 @@ def __init__(self, http_port: int = args.http_port,
-    def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio: bool = False):
+            FastAPIRoute(r"/v1/models/{model_name}", dataplane.model_ready),
+            FastAPIRoute(r"/v1/models/{model_name}:predict", dataplane.infer, methods=["POST"]),
+            FastAPIRoute(r"/v1/models/{model_name}:explain", dataplane.infer, methods=["POST"]),
```
should be calling `explain`

yuzisun commented on python/kserve/kserve/model_server.py:
```diff
@@ -132,30 +112,16 @@ def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio:
-
-        logging.info(f"Setting max asyncio worker threads as {self.max_asyncio_workers}")
-        asyncio.get_event_loop().set_default_executor(
-            concurrent.futures.ThreadPoolExecutor(max_workers=self.max_asyncio_workers))
```
Do we still need this asyncio worker code for fastAPI ? I think it is not tornado specific.

imfing committed:
Add v2 routes with FastAPIRoute

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Extract errors and add exception handler for ModelServer

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add metrics route and handler

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add kserve version in __init__.py

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Remove tornado handlers

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Move ModelNotReady to errors.py

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Remove unused arguments

- max_buffer_size
- max_asyncio_workers

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Clean up validate_enable_latency_logging
by using strtobool

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Remove tests for validate_enable_latency_logging

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Improve error response messages

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Update handlers

- fix imports for model_repository_extension.py
- raise ModelNotFound exception in model_ready handler

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add model ready and list handler

- replace model_metadata with model_list_handler for /v1/models
- use ModelRepositoryExtension for load and unload

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix misspelling

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use FastAPI TestClient for testing

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

sukumargaonkar committed:
Setup up single source versioning for python sdk

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add V1Endpoints and update tests

- switch to use v1_endpoints in FastAPI routes
- make TestModel and TestTFHttpServer pass

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix TestRayServer

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix model not ready test

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix cloud event tests

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use pytest asyncio and lowercase header key

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Set default header to json

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Update tests

- fixed TestModel
- fixed load/unload tests

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add V2Endpoints and data models

- in v1, rename `infer` to `predict`

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Improve Swagger UI support

- disable redoc and add flag for swagger ui
- add example for InferenceRequest model

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

sukumargaonkar committed:
Fix tests

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

alexagriffith commented on python/VERSION:
```diff
@@ -0,0 +1 @@
+0.9.0
```
ide complaining about no line at eof

alexagriffith replied:
also do you think we could name this a little differently ? like `python_sdk/VERSION` ? python could be confusing

imfing committed:
Revert "Add kserve version in __init__.py"

This reverts commit a0204f8bcd7cc849ae55127355d07e9d5bd4eb92.

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix the version number in setup.py

by using the pathlib to get the __file__ path instead of the cwd

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix the version file in dockerfile

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix linting issue

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix other lint issues

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix lint issue

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use version from VERSION file

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use `Dict` for typing

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix `Dict` typing

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix `List` typing

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Support v2 load and unload

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix test_model_repository_extension.py

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix dataplane

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add orjson in requirements.txt

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use httpx.AsyncClient to replace tornado

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Remove tornado

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add httpx in requirements.txt

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Update doc strings for some functions

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use orjson for decoding

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use async for liveness handler

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Added a few comments

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Unify infer and predict in dataplane

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Get rid of json

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use ORJSONResponse by default in FastAPI application

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix orjson exception match

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add more comments

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add handler for NotImplementedError

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use orjson for InferenceRequest deserialization

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add schema examples

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix v2 live and ready endpoints

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Use asyncio to fix start() issue

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix missing asyncio import

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add missing asyncio to start model server

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Set default host to 0.0.0.0

to fix raw deployment readiness issue

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun commented on python/aixexplainer/setup.py:
```diff
@@ -31,7 +36,8 @@
     install_requires=[
-        "kserve",
+        f"kserve>={version}",
+        "argparse >= 1.4.0",
```
Do we need argparse ? I think it is part of the kserve dependencies

yuzisun commented:
Let’s add the breaking changes to the release note on PR description. I think the only api breaking change is the need to add asyncio.run to start the model server right ?

yuzisun commented on python/kserve/kserve/handlers/model_repository_extension.py:
```diff
@@ -0,0 +1,88 @@
+import inspect
```
missing license file

yuzisun commented on python/kserve/kserve/handlers/v1_endpoints.py:
```diff
@@ -0,0 +1,89 @@
+# Copyright 2021 The KServe Authors.
```
```suggestion
# Copyright 2022 The KServe Authors.
```

yuzisun commented on python/kserve/kserve/handlers/v2_endpoints.py:
```diff
@@ -0,0 +1,116 @@
+# Copyright 2021 The KServe Authors.
```
```suggestion
# Copyright 2022 The KServe Authors.
```

yuzisun commented on python/kserve/kserve/handlers/v2_endpoints.py:
```diff
@@ -0,0 +1,116 @@
+            model_name (str): Model name.
+
+        Returns:
+            Dict: {"name": model_name, "load": True}
```
This also raise the exception what `model_repository_extension` throws

yuzisun commented on python/aiffairness/aifserver/__main__.py:
```diff
@@ -71,4 +71,4 @@
         unprivileged_groups=args.unprivileged_groups
     )
     model.load()
-    kserve.ModelServer().start([model], nest_asyncio=True)
```
looks like `nest_asyncio` is still required for explainers as it is using the asyncio loop here
https://github.com/kserve/kserve/blob/master/python/aixexplainer/aixserver/model.py#L51

yuzisun committed:
add back nest_asyncio for explainers

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

imfing committed:
Use async for load and unload

for model_repository_extension.py and update some comments

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Get rid of DeprecationWarning from cloudevents

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Fix test_model_repository_extension.py with async

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Add license

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

imfing committed:
Update comments and docstrings

Signed-off-by: Xin Fu <xfu83@bloomberg.net>

yuzisun commented:
Awesome work!! @xfu83 and @sukumargaonkar 

/lgtm
/approve
